### DISCUSSION TITLE
Best way to load a large number of images from URLs

### DISCUSSION THREAD

**Discussion by christosk92** [AvaloniaUI/Avalonia#9878] [Q&A] [answered]
Basically I have a For You Page with a bunch of album art from a music service (CDN url).

I noticed unlike UWP/WPF, ``<Image/>`` only accepts local paths.  

What would be the best way to load a bunch of these images in parallel in an efficient way? I tried making a custom converter, downloading the image to a stream and then loading it, but this consumes massive amounts of Memory and is slow. Takes almost 5 seconds or more to load all the images. 

For reference, loading roughly 50 images with a size of 180x180 pretty much loads instantly (<200 ms) in WinUI/UWP/WPF.

**Comment by maxkatz6** [comment, marked as answer]
> I noticed unlike UWP/WPF, <Image/> only accepts local paths.

Image.Source accepts IImage, which usually is an instance of Bitmap.
Bitmap can be created from a stream.

> I tried making a custom converter, downloading the image to a stream and then loading it, but this consumes massive amounts of Memory and is slow. 

It really depends on how exactly you are downloading these images. Is it async, is it in parallel...

Before I used this image subclass with some HttpClient helpers https://gist.github.com/maxkatz6/bc6e25db4323ede3b9a631cd7511cdb5
You might notice `Bitmap.DecodeToHeight` part - it is mostly needed if you have large original images that needs to be resized to a smaller size while decoding. In your case it's probably not needed.

**Reply by timunie** [reply]
If you need async image loading you can also bind to `Tasky<IImage>`, see: https://docs.avaloniaui.net/docs/data-binding/binding-to-tasks-and-observables#example-2-binding-to-a-task 

In my use-case I load the images like shown above and in addition I have a caching, for example: https://learn.microsoft.com/en-us/dotnet/core/extensions/caching .

**Reply by tuyen-vuduc** [reply]
There is a NuGet package to save us time.

```
<PackageReference Include="AsyncImageLoader.Avalonia" Version="3.2.1" />
```

ref: https://stackoverflow.com/questions/67536123/in-avaloniaui-how-to-display-an-image-from-a-web-url

**Comment by christosk92** [comment]
Wow I did not think of making a subclass of Image because they are sealed in UWP. 

That sounds like a good idea, I will try that. Thank you
